### PR TITLE
Do not cast residue ids/numbers in `Topology.from_openmm`

### LINKED ISSUE #1953: `Topology.from_openmm` casts residue number strings to integers
**Describe the bug**

In `Topology.from_openmm`, residue numbers (which are strings) [are cast to integers](https://github.com/openforcefield/openff-toolkit/blame/6aa989ed54c0bdd9a9a504f008d0719753541246/openff/toolkit/topology/topology.py#L1524-L1526), which are later not converted back to strings. This causes problems when interoperating with tools that follow the [conventional behavior of OpenMM objects.
](http://docs.openmm.org/development/api-python/generated/openmm.app.topology.Topology.html#openmm.app.topology.Topology.addResidue)
`Topology.from_pdb` exhibits what I'm calling the correct behavior (keeping residue numbers as strings), which when sent to OpenMM can be read by downstream tools.

**To Reproduce**

```python
import openmm.app
import parmed

from openff.toolkit import Topology
from openff.toolkit.utils import get_data_file_path


ala_path = get_data_file_path("proteins/MainChain_ALA_ALA.pdb")

openmm_topology = openmm.app.PDBFile(ala_path).topology
openff_topology = Topology.from_pdb(ala_path)
openff_molecule = openff_topology.molecule(0)

case1 = {type(residue.id) for residue in openmm_topology.residues()}
case2 = {type(residue.residue_number) for residue in openff_topology.residues}
case3 = {
    type(residue.residue_number)
    for residue in Topology.from_openmm(
        openmm_topology=openmm_topology,
        unique_molecules=[openff_molecule],
    ).residues
}
case4 = {
    type(residue.id)
    for residue in Topology.from_openmm(
        openmm_topology=openmm_topology,
        unique_molecules=[openff_molecule],
    )
    .to_openmm()
    .residues()
}

print(f"openmm.app.PDBFile: {case1}")
print(f"openff.toolkit.Topology.from_pdb: {case2}")
print(f"openff.toolkit.Topology.from_openmm: {case3}")
print(f"openff.toolkit.Topology.from_openmm.to_openmm: {case4}")

parmed.openmm.load_topology(openmm_topology)
parmed.openmm.load_topology(openff_topology.to_openmm())

parmed.openmm.load_topology(
    Topology.from_openmm(
        openmm_topology=openmm_topology,
        unique_molecules=[openff_molecule],
    ).to_openmm()
)
```

**Output**

```shell
$ python metadata-type-bug.py                                                                                            10:12:56  ☁  6aa989ed ☂
openmm.app.PDBFile: {<class 'str'>}
openff.toolkit.Topology.from_pdb: {<class 'str'>}
openff.toolkit.Topology.from_openmm: {<class 'int'>}
openff.toolkit.Topology.from_openmm.to_openmm: {<class 'int'>}
Traceback (most recent call last):
  File "/Users/mattthompson/software/openff-toolkit/metadata-type-bug.py", line 41, in <module>
    parmed.openmm.load_topology(
  File "/Users/mattthompson/micromamba/envs/openeye-examples/lib/python3.10/site-packages/parmed/utils/decorators.py", line 18, in new_fcn
    return fcn(*args, **kwargs)
  File "/Users/mattthompson/micromamba/envs/openeye-examples/lib/python3.10/site-packages/parmed/openmm/topsystem.py", line 91, in load_topology
    resnum = int(r.id) if r.id.lstrip('-').isdigit() else r.index
AttributeError: 'int' object has no attribute 'lstrip'
```

**Computing environment (please complete the following information):**
 - Operating system: macOS
 - Output of running `conda list`: https://gist.github.com/mattwthompson/38d94cfbaeb6e0193921595dbb6e948d

**Additional context**

Related to #1722

### PR DESCRIPTION
**Pull request by mattwthompson**
Fixes #1953

- [x] Tag issue being addressed
- [x] Add [tests](https://github.com/openforcefield/openff-toolkit/tree/main/openff/toolkit/_tests)
- [ ] Update docstrings/[documentation](https://github.com/openforcefield/openff-toolkit/tree/main/docs), if applicable
- [x] [Lint](https://open-forcefield-toolkit.readthedocs.io/en/latest/developing.html#style-guide) codebase
- [ ] Update [changelog](https://github.com/openforcefield/openff-toolkit/blob/main/docs/releasehistory.md)

### CODE REVIEW

File: openff/toolkit/_tests/test_topology.py [hunk 1]
```diff
@@ -754,7 +754,9 @@ def test_to_from_openmm_hierarchy_metadata(self):
                 roundtrip = roundtrip_atom.metadata["residue_number"]
                 assert original == roundtrip
             else:
-                assert roundtrip_atom.metadata["residue_number"] == 0
+                # OpenFF lets residue numbers be str or int, but this is str in OpenMM's
+                # representation, so faithful round-trip is not practical
```
Comment (j-wags) [inline openff/toolkit/_tests/test_topology.py:758]: (blocking) Good explanation but confusing when read outside the context of this PR, let's make sure to delete this comment before merging

**Review by j-wags — APPROVED** [review]
Great. I think this makes sense. For a while I churned in my head thinking about if this would break some existing code, but ultimately I don't think it matters because retaining the data type from OpenMM on IMPORT and casting on EXPORT is the right thing to do in general. 

Please update the releasenotes before merging!

**Comment by mattwthompson** [comment]
> casting on EXPORT is the right thing to do in general 

That's not quite what's happening here - this change just prevents the import cast

```python
In [1]: from openff.toolkit import Molecule

In [2]: molecule = Molecule.from_smiles("CC")

In [3]: for atom in molecule.atoms:
   ...:     atom.metadata["residue_number"] = 4
   ...:

In [4]: openmm_topology = molecule.to_topology().to_openmm()

In [6]: {type(residue.id) for residue in openmm_topology.residues()}
Out[6]: {int}

**Comment by j-wags** [comment]
Ah yes, apologies for the confusion, that's what I meant.

**Comment by mattwthompson** [comment]
Great - let's keep this change as-is and split the discussion of the output type elsewhere

**Pull request merged**